src/cider/common/interpreters/AggregationHashTable.h

```diff
@@ -0,0 +1,245 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <common/Arena.h>
+#include <common/hashtable/FixedHashMap.h>
+#include <common/hashtable/HashMap.h>
+#include <type/data/sqltypes.h>
+
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <type_traits>
+#include <unordered_set>
+
+#include "cider/CiderException.h"
+
+namespace Cider {
+
+using AggregateDataPtr = int8_t*;
+
+using AggregatedDataWithUInt8Key =
+    FixedImplicitZeroHashMapWithCalculatedSize<uint8_t, AggregateDataPtr>;
+using AggregatedDataWithUInt16Key = FixedImplicitZeroHashMap<uint16_t, AggregateDataPtr>;
+
+template <typename Base>
+struct AggregationDataWithNullKey : public Base {
+  using Base::Base;
+
+  bool& hasNullKeyData() { return has_null_key_data; }
+  AggregateDataPtr& getNullKeyData() { return null_key_data; }
+  bool hasNullKeyData() const { return has_null_key_data; }
+  const AggregateDataPtr& getNullKeyData() const { return null_key_data; }
+  size_t size() const { return Base::size() + (has_null_key_data ? 1 : 0); }
+  bool empty() const { return Base::empty() && !has_null_key_data; }
+  void clear() {
+    Base::clear();
+    has_null_key_data = false;
+  }
+  void clearAndShrink() {
+    Base::clearAndShrink();
+    has_null_key_data = false;
+  }
+
+ private:
+  bool has_null_key_data = false;
+  AggregateDataPtr null_key_data = nullptr;
+};
+
+template <typename... Types>
+using HashTableWithNullKey = AggregationDataWithNullKey<HashMapTable<Types...>>;
+
+using AggregatedDataWithNullableUInt8Key =
+    AggregationDataWithNullKey<AggregatedDataWithUInt8Key>;
+using AggregatedDataWithNullableUInt16Key =
+    AggregationDataWithNullKey<AggregatedDataWithUInt16Key>;
+
+class AggregationHashTable;
+HashTableAllocator allocator;
+
+struct AggregationMethod : private boost::noncopyable {
+  enum class Type {
+    EMPTY = 0,
+    INT8 = 1,
+    INT16 = 2,
+    SERIALIZED = 3,
+    without_key,
+  };
+  Type type = Type::EMPTY;
+
+  AggregationMethod() {}
+
+  ~AggregationMethod();
+};
+
+class AggKey {
+ public:
+  bool is_null;
+  int8_t* addr;
+  uint32_t len;
+
+  AggKey(bool is_null, int8_t* addr, uint32_t len) {
+    this->is_null = is_null;
+    this->addr = addr;
+    this->len = len;
+  }
+
+  bool isNull() const { return is_null; }
+
+  int8_t* getAddr() const { return addr; }
+
+  uint32_t getLen() const { return len; }
+
+  const bool operator==(const AggKey& key) const {
+    return this->is_null == key.is_null &&
+           std::memcmp(this->addr, key.addr, std::min(this->len, key.len));
+  }
+};
+
+class AggregationHashTable final {

```

Comment:
Why need this class. will this be visible to hash table user or just cider library internal use. 
if visible to hash table user, where should it be on the class diagram?